programs/test/udp_proxy.c

```diff
@@ -560,14 +662,30 @@ int main( int argc, char *argv[] )
         nb_fds = listen_fd.fd;
     ++nb_fds;
 
+    if( opt.merge > 0 )
+    {
+        outbuf[0].ctx = &server_fd;
+        outbuf[0].description = "S <- C";
+        outbuf[0].num_datagrams = 0;
+        outbuf[0].len = 0;
+
+        outbuf[1].ctx = &client_fd;
+        outbuf[1].description = "S -> C";
+        outbuf[1].num_datagrams = 0;
+        outbuf[1].len = 0;
+    }
+
     while( 1 )
     {
         FD_ZERO( &read_fds );
         FD_SET( server_fd.fd, &read_fds );
         FD_SET( client_fd.fd, &read_fds );
         FD_SET( listen_fd.fd, &read_fds );
 
-        if( ( ret = select( nb_fds, &read_fds, NULL, NULL, NULL ) ) <= 0 )
+        ctx_buffer_check( &outbuf[0] );
+        ctx_buffer_check( &outbuf[1] );
+
+        if( ( ret = select( nb_fds, &read_fds, NULL, NULL, &tm ) ) < 0 )
```

Comment:
Here `select` is actually used to poll on multiple file descriptors simultaneously, something that `mbedtls_net_poll` isn't able to do.

Reply:
But I'm not convinced the simultaneity is critical, it seems to me we could replace each of the subsequent `FD_ISSET()` with a direct call to `mbedtls_net_poll()`. Anyway, that's tangential and not important for acceptance of the PR.

Reply:
@mpg Sorry, I mistakenly remembered that this was a blocking call to `select`, but it's not - and even if it had been, since we're in a loop here anyway we can use the non-blocking version of `net_poll` sequentially for client and server.

Reply:
I just realized that this is a busy-wait loop. Please don't do that, even on a host-only program. The use of `select` is justified here, but pass an infinite timeout.

Reply:
@gilles-peskine-arm @mpg If we want to pass an infinite timeout, then we have to stick to `select`, though, and cannot use `mbedtls_net_poll`.

Reply:
@gilles-peskine-arm As far as I see, we're busy-waiting in `ssl_server2` and `ssl_client2`, too.

Reply:
I think the timeout should not be infinite, but computed as the minimum of the "wait to see if there are other packets to pack with the current" times of the two buffer contexts.

(Btw I agree the use of `select()` is justified then.)